### PR TITLE
feat: integrate github oauth

### PR DESCRIPTION
**Pull request by definitelynobody**

### CODE REVIEW

**Comment by rvolosatovs** [comment]
Please rebase on latest `main` :)

File: crates/auth/src/providers/auth0/mod.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+impl std::error::Error for ValidateError {}
+
+#[derive(Deserialize)]
+pub(crate) struct Auth0User {
+    #[serde(rename = "login")]
+    pub username: String,
+    pub id: u64,
+}
+
+//////////////////////// TODO
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:54]: What is this for?

File: crates/auth/src/providers/auth0/mod.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+use std::{fmt, io};
+
+use axum::http::header::{AUTHORIZATION, USER_AGENT};
+use axum::http::status::InvalidStatusCode;
+use axum::http::StatusCode;
+use oauth2::basic::BasicClient;
+use oauth2::{AuthUrl, ClientId, RedirectUrl, TokenUrl};
+use serde::Deserialize;
+
+pub const LOGIN_URI: &str = "/auth/auth0";
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:17]:
```suggestion
pub const LOGIN_URI: &str = "/login";
```

File: crates/auth/src/providers/auth0/mod.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+pub const LOGOUT_URI: &str = "/auth/auth0/logout";
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:18]:
```suggestion
pub const LOGOUT_URI: &str = "/logout";
```

File: crates/auth/src/providers/auth0/mod.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+// pub const AUTHORIZED_URI: &str = "/auth/auth0/authorized";
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:19]:
```suggestion
```

File: crates/auth/src/providers/auth0/mod.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+//////////////////////// TODO
+pub(crate) async fn validate(session: &Session) -> Result<Auth0User, ValidateError> {
+    #[derive(Deserialize)]
+    struct Error {
+        message: String,
+    }
+
+    let res = ureq::get("https://api.github.com/user")
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:61]: This should be doing a `UserInfo` request using the configured OIDC client https://github.com/profianinc/drawbridge/blob/bb43cb593f04030b376143bfe55476f06da025fc/crates/app/src/auth/oidc.rs#L94-L113

File: crates/auth/src/providers/auth0/mod.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+
+#[derive(Clone)]
+pub struct OAuthClient(pub BasicClient);
+
+impl OAuthClient {
+    pub fn new(host: &str, auth0_host: &str, client_id: String) -> OAuthClient {
+        let auth_url = format!("https://{}/authorize", auth0_host);
+        let token_url = format!("https://{}/oauth/token", auth0_host);
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:85]: What are these for? Why are `/login`, `/logout` and `/authorized` not enough?

File: crates/auth/src/providers/auth0/mod.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+        OAuthClient(
+            BasicClient::new(
+                ClientId::new(client_id),
+                None,
+                AuthUrl::new(auth_url).unwrap(),
+                Some(TokenUrl::new(token_url).unwrap()),
+            )
+            .set_redirect_uri(
+                RedirectUrl::new(format!("http://{}{}", host, AUTHORIZED_URI)).unwrap(),
```
Comment (rvolosatovs) [inline crates/auth/src/providers/auth0/mod.rs:95]: We should host this on https and never redirect to `http`

File: crates/auth/src/providers/github/mod.rs [hunk 8]
```diff
@@ -14,11 +14,12 @@ use oauth2::basic::BasicClient;
 use oauth2::{AuthUrl, ClientId, ClientSecret, RedirectUrl, TokenUrl};
 use serde::Deserialize;
```
Comment (rvolosatovs) [inline crates/auth/src/providers/github/mod.rs:16]: Github provider can be completely removed

File: crates/auth/src/providers/mod.rs [hunk 9]
```diff
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 
 pub mod github;
```
Comment (rvolosatovs) [inline crates/auth/src/providers/mod.rs:4]:
```suggestion
```

File: crates/auth/src/providers/mod.rs [hunk 10]
```diff
@@ -10,6 +11,7 @@ use serde::{Deserialize, Serialize};
 #[derive(Debug, Serialize, Deserialize)]
 pub enum Provider {
     GitHub,
```
Comment (rvolosatovs) [inline crates/auth/src/providers/mod.rs:13]:
```suggestion
```

**Review by rvolosatovs — CHANGES_REQUESTED** [review]
Please rebase on latest `main` and follow the description in https://github.com/profianinc/benefice/issues/6, the process is explained there step-by-step with references to relevant documentation and copy-pastable code.
I do not understand how this relate to the issue at hand and I am happy to jump on a call to discuss what this should do.

**Comment by npmccallum** [comment]
@definitelynobody This is blocking other important features and needs to be pushed over the line ASAP.

**Pull request closed without merging**